### ISSUE TITLE
Training speed is very slow on multi gpus

### ISSUE THREAD

**Issue by simon5u** [ShiqiYu/libfacedetection.train#3]
Hi, I am trying to repeat your experiment by running your train.py in one gpu and four gpus, it turns out 4 gpus require longer time than one gpu, 150+ hours versus 60+ hours. I am using 2080Ti on ubuntu 16, cuda 10.1, nvidia 430, pytorch 1.2.0. Did you encounter the same problem on your side? Thanks.

**Comment by AmazingRoad** [comment]
what is the usage rate of your gpus?

**Comment by ShiqiYu** [comment]
@simon5u When I train the model using multiple GPU, the speed is the same. The reason may be that the model is too simple.

**Comment by simon5u** [comment]
@ShiqiYu Thanks for the reply. I think the problem is from data.py -> def detection_collate(batch). May be can refer to others who train face detector using widerface. Anyway, thanks for the sharing.

**Comment by ShiqiYu** [comment]
> @ShiqiYu Thanks for the reply. I think the problem is from data.py -> def detection_collate(batch). May be can refer to others who train face detector using widerface. Anyway, thanks for the sharing.

Do you have any suggestions to optimize the source code to speedup?

**Comment by DuckJ** [comment]
@ShiqiYu [https://github.com/ShiqiYu/libfacedetection.train/blob/master/tasks/task1/train.py#L127](https://github.com/ShiqiYu/libfacedetection.train/blob/master/tasks/task1/train.py#L127)
这句是不是应该移到for循环外面，这样会提速吧

**Comment by simon5u** [comment]
@ShiqiYu i tried the following change and it is working better now:-
        for iteration in range(args.resume_epoch, epoch_size):
            if iteration % epoch_size == 0:
                # create batch iterator
                batch_iterator = iter(data.DataLoader(dataset, batch_size, shuffle=True,
                                                  num_workers=num_workers, collate_fn=detection_collate))

thanks to @DuckJ !

**Comment by ShiqiYu** [comment]
> @ShiqiYu i tried the following change and it is working better now:-
> for iteration in range(args.resume_epoch, epoch_size):
> if iteration % epoch_size == 0:
> # create batch iterator
> batch_iterator = iter(data.DataLoader(dataset, batch_size, shuffle=True,
> num_workers=num_workers, collate_fn=detection_collate))
> 
> thanks to @DuckJ !

Dear @simon5u I updated the code with your suggestion. The training is 10x faster when I set num_workers=8 and use only 1 GPU. But I still cannot train faster with multiple GPUs

**Comment by ShiqiYu** [comment]
> @ShiqiYu https://github.com/ShiqiYu/libfacedetection.train/blob/master/tasks/task1/train.py#L127
> 这句是不是应该移到for循环外面，这样会提速吧

Done! Thank you!
